### DISCUSSION TITLE
How to make tunnel persistent at boot?

### DISCUSSION THREAD

**Discussion by msanaullahsahar** [loophole/cli#150] [General]
Hi, I am using loophole to access my test website hosted on my raspberry pi. I am using these two commands to expose my website.
`cd home/pi/loophole`
`./loophole http 80 --hostname mytestwebsite`
This works fine and i am able to access my website at the generated url until my raspberry pi does not get restart. So how can i make this persistent to eliminate the need of running this code again and again at every boot?

**Comment by 0x7f** [comment]
Hi @msanaullahsahar glad you are happily using loophole :)

The loophole cli itself is not a service. As you mention, it is just a binary for starting tunnels ad hoc. So when you want some binary to be executed with every boot of your raspi, you need to set up a system service. I guess you use something like raspbian on you pi which is using `systemd` for managing the services. There are couple of resources in the internet on how to write systemd service units (e.g. [here](https://www.freedesktop.org/software/systemd/man/systemd.service.html), [here](http://tuxgraphics.org/npa/systemd-scripts/), [here](https://carlpaton.github.io/2020/10/raspberry-create-a-service/) and [here](https://domoticproject.com/creating-raspberry-pi-service/)).

In your case the file could look roughly like this (did not test the file):
```ini
[Unit]
Description=Loophole
After=network-online.target

[Service]
ExecStart=./loophole http 80 --hostname mytestwebsite
WorkingDirectory=/home/pi/loophole
StandardOutput=inherit
StandardError=inherit
Restart=always
User=pi

[Install]
WantedBy=multi-user.target
```

Save this file as e.g. `loophole.service` and store it in `/etc/systemd/system`.  You can start the service via `systemctl start loophole.service`. Use `systemctl status loophole.service` to check whether it successfully started. Finally, use `systemctl enable loophole.service` to make sure the service is started on every boot.

**Reply by msanaullahsahar** [reply]
Thank you @0x7f for quick answer. I did the same however the service returned the following error.

Failed to start loophole.service as the unit loophole.service has a bad unit file setting. `systemctl status loophole.service` displayed the following status about this service.

```
● loophole.service - Loophole
     Loaded: bad-setting (Reason: Unit loophole.service has a bad unit file setting.)
     Active: inactive (dead)

Nov 27 01:55:49 raspberrypi systemd[1]: /etc/systemd/system/loophole.service:6: Neither a valid executable name nor an absolute path: ./loophole
Nov 27 01:55:49 raspberrypi systemd[1]: loophole.service: Unit configuration has fatal error, unit will not be started.
```

Does this error means we can not execute the file `loophole` using `./`? Are there other ways to run this without `./`? I have tried with `bash loophole` and `sh loophole` but no luck. Any solution is appreciated and again thanks for your time.

**Comment by msanaullahsahar** [comment]
Ok. I replaced `ExecStart=./loophole http 80 --hostname mytestwebsite` with `ExecStart=/home/pi/loophole/loophole http 80 --hostname mytestwebsite`
Now i can start the service using `systemctl start loophole.service` but the service does not start itself on boot. Offcourse i have used the following commands after every change that I made.
```
sudo systemctl daemon-reload
sudo systemctl start loophole.service
sudo systemctl enable loophole.service
```
I am stuck due to my limited knowledge of linux.

**Reply by Morishiri** [reply]
Can you check if `systemctl status loophole.service` gives more details this time? Is there a startup error or the service is not even started?

Additionally some time ago I was experiencing issues with services when they were started from user's home directory, try moving to file to some other location, e.g. `/usr/local/bin` or `/opt`, changing the path in service description and trying again?

**Comment by jmeri** [comment]
Or you could just add the line

@reboot  /home/pi/loophole/loophole http 80 --hostname mytestwebsite > /dev/null 2>&1

to your crontab.
